File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第4章：对比度增强/4.4/gamma.py
# -*- coding: utf-8 -*-
import cv2
import numpy as np
import sys
#主函数
if __name__ =="__main__":
    if len(sys.argv) > 1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python gamma.py imageFile"
    MAX_VALUE = 200
    value = 40
    segValue = float(value)
    #伽马调整需要先将图像归一化
    image_0_1 = image/255.0
    #伽马调整后的图像显示窗口
    cv2.namedWindow("gamma_contrast",cv2.WND_PROP_AUTOSIZE)
    #调整 gamma 值，观察图像的变换
    def callback_contrast(_value):
        gamma = float(_value)/segValue
        contrastImage = np.power(image_0_1,gamma)
        cv2.imshow("gamma_contrast",contrastImage)
        #保存伽马调整的结果
        contrastImage*=255
        contrastImage = np.round(contrastImage)
        contrastImage = contrastImage.astype(np.uint8)
        cv2.imwrite("gamma.jpg",contrastImage)
    callback_contrast(value)
    cv2.createTrackbar("value","gamma_contrast",value,MAX_VALUE,callback_contrast)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第4章：对比度增强/4.4/gamma.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
#include<string>
#include<iostream>
cv::Mat image;//输入图像
cv::Mat image_0_1;//归一化输入图像
cv::Mat contrastImage;//输出图像
//图像的宽高
int height;
int width;
int value = 16;
float segValue = 40.0;
const int MAX_VALUE = 120;
std::string contrastWindow = "对比度";
void callback_value(int, void*)
{
	//这里 gamma 的取值范围为 [0，3]
	float gamma = float(value) / segValue;
	//调整对比度
	for (int r = 0; r < height; r++)
	{
		for (int c = 0; c < width; c++)
		{
			float pixel = powf(image_0_1.at<float>(r, c), gamma);
			contrastImage.at<float>(r, c) = pixel;
		}
	}
	//或者直接采用 OpenCV 提供的 pow函数
	//cv::pow(image_0_1, gamma, contrastImage);
	//显示调整对比度后的效果
	cv::imshow(contrastWindow, contrastImage);
	/*保存结果*/
	contrastImage.convertTo(contrastImage, CV_8UC1, 255, 0);
	cv::imwrite("contrastImage.jpg", contrastImage);
}
int main(int argc, char*argv[])
{
	//输入图片
	image = cv::imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
	{
		std::cout << "没有图片" << std::endl;
		return -1;
	}
	//图片归一化，每一个灰度值都除以 255
	image.convertTo(image_0_1, CV_32FC1, 1.0 / 255.0, 0);
	//显示原图
	cv::imshow("原图", image_0_1);
	//图像的宽高
	height = image.rows;
	width = image.cols;
	contrastImage = cv::Mat::zeros(image.size(), CV_32FC1);
	cv::namedWindow(contrastWindow, CV_WINDOW_AUTOSIZE);
	cv::createTrackbar("value", contrastWindow, &value, MAX_VALUE, callback_value);
	callback_value(0, 0);
	cv::waitKey(0);
	return 0;
}